clang/runtime/dpct-rt/include/dpct/dpl_extras/dpcpp_extensions.h

```diff
@@ -373,34 +374,108 @@ template <typename T, int VALUES_PER_THREAD> class exchange {
 
   exchange(uint8_t *local_memory) : _local_memory(local_memory) {}
 
-  /// Rearrange elements from rank order to blocked order
-  template <typename Item>
-  __dpct_inline__ void
-  scatter_to_blocked(Item item, T (&keys)[VALUES_PER_THREAD],
-                     int (&ranks)[VALUES_PER_THREAD]) {
+  // TODO: Investigate if padding is required for performance,
+  // and if specializations are required for specific target hardware.
+  static size_t adjust_by_padding(size_t offset) {
+
+    if constexpr (INSERT_PADDING) {
+      offset = detail::shr_add(offset, LOG_LOCAL_MEMORY_BANKS, offset);
+    }
+    return offset;
+  }
+
+  struct get_striped_offset {
+    template <typename Item> size_t operator()(Item item, size_t i) {
+      size_t offset = item.get_local_id(0) * VALUES_PER_THREAD + i;
+      return adjust_by_padding(offset);
+    }
+  };
+
+  struct get_blocked_offset {
+    template <typename Item> size_t operator()(Item item, size_t i) {
+      size_t offset = i * item.get_local_range(2) * item.get_local_range(1) *
```

Comment:
This looks like a striped offset to me. Adjacent work items are accessing adjacent elements in memory which is a striped pattern. Should this functor be called `get_striped_offset`?